### PR TITLE
devour component

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
makes it a component

### CODE REVIEW

**Comment by Rohesie** [comment]
```
[19:34:02] Runtime in component_devour.dm, line 138: Cannot execute null.forceMove().
proc name: remove devoured (/datum/component/ability/devour/proc/remove_devoured)
usr: CKEY/(Mature Larva (204))
usr.loc: (North Central Barrens (117, 121, 2))
src: /datum/component/ability/devou... (/datum/component/ability/devour)
call stack:
/datum/component/ability/devou... (/datum/component/ability/devour): remove devoured(the Mature Larva (204) (/mob/living/carbon/xenomorph/larva))
/datum/callback (/datum/callback): InvokeAsync(the Mature Larva (204) (/mob/living/carbon/xenomorph/larva))
the Mature Larva (204) (/mob/living/carbon/xenomorph/larva): SendSignal("xenomorph_evolve", /list (/list))
the Mature Larva (204) (/mob/living/carbon/xenomorph/larva): do evolve(null, null)
the Mature Larva (204) (/mob/living/carbon/xenomorph/larva): Evolve()
```
```
[19:33:18] Runtime in component_devour.dm, line 35: Cannot execute null.remove action().
proc name: UnregisterFromParent (/datum/component/ability/devour/UnregisterFromParent)
usr: *no key*/(Mature Larva (34))
usr.loc: (Central Caves (74, 160, 2))
src: /datum/component/ability/devou... (/datum/component/ability/devour)
call stack:
/datum/component/ability/devou... (/datum/component/ability/devour): UnregisterFromParent()
/datum/component/ability/devou... (/datum/component/ability/devour): RemoveFromParent()
/datum/component/ability/devou... (/datum/component/ability/devour): Destroy(0, 1)
/datum/component/ability/devou... (/datum/component/ability/devour): Destroy(0, 1)
qdel(/datum/component/ability/devou... (/datum/component/ability/devour), 0, 1)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Destroy(0)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Destroy(0)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Destroy(0)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Destroy(0)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Destroy(0)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Destroy(0)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Destroy(0)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Destroy(0)
qdel(the Mature Larva (34) (/mob/living/carbon/xenomorph/larva), 0)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): do evolve(null, null)
the Mature Larva (34) (/mob/living/carbon/xenomorph/larva): Evolve()
```

File: code/datums/components/abilities/component_devour.dm [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+		stack_trace("[stowaway] found in [P]'s contents. It shouldn't have ended there.")
+
+// Regurgitate
+/datum/action/ability/regurgitate
+	name = "Regurgitate"
+	action_icon_state = "regurgitate"
+	//mechanics_text = "Vomit whatever you have devoured."
+
+/datum/action/ability/regurgitate/action_activate()
+	SEND_SIGNAL(owner, COMSIG_XENOABILITY_REGURGITATE)
```
Comment (ninjanomnom) [inline code/datums/components/abilities/component_devour.dm]: Use a generic activate signal

File: code/datums/components/abilities/component_devour.dm [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+		COMSIG_XENOMORPH_DEEVOLVE
+	))
+
+/datum/component/ability/devour/proc/devour(datum/source)
+	if(devoured)
+		to_chat(parent, "<span class='warning'>You already have something in your belly, there's no way that will fit.</span>")
+		return NONE
+	var/turf/T = get_turf(parent)
+	for(var/i in T)
+		var/atom/A = i
```
Comment (ninjanomnom) [inline code/datums/components/abilities/component_devour.dm]: Too many single character names

File: code/datums/components/abilities/component_devour.dm [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		if(A == parent)
+			continue
+		if(A.density)
+			to_chat(parent, "<span class='danger'>You can't do that here!</span>")
+			return NONE
+
+	var/mob/P = parent
```
Comment (ninjanomnom) [inline code/datums/components/abilities/component_devour.dm]: Ditto

File: code/datums/components/abilities/component_devour.dm [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	var/atom/movable/grab = P.pulling
+	var/mob/as_mob = grab
```
Comment (ninjanomnom) [inline code/datums/components/abilities/component_devour.dm]: Recast this where you use it

File: code/datums/components/abilities/component_devour.dm [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	"<span class='danger'>You start to devour [grab]!</span>", null, 5)
+
+	if(!do_after(P, 5 SECONDS, FALSE, grab, BUSY_ICON_DANGER, extra_checks = CALLBACK(src, .proc/can_devour_grabbed, grab)))
+		to_chat(P, "<span class='warning'>You stop devouring \the [grab]. \He probably tasted gross anyways.</span>")
+		return NONE
+
+	P.visible_message("<span class='warning'>[P] devours [grab]!</span>", \
+	"<span class='warning'>You devour [grab]!</span>", null, 5)
+
+	grab.moveToNullspace()
```
Comment (ninjanomnom) [inline code/datums/components/abilities/component_devour.dm]: Don't have this middle step, mobs should not be in nullspace ever

File: code/datums/components/abilities/component_devour.dm [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+			var/mob/living/as_living = grab
+			as_living.knock_down(360)
+			as_living.blind_eyes(1)
+
+	SEND_SIGNAL(grab, COMSIG_MOVABLE_DEVOURED, P)
+
+	return COMSIG_GRAB_SUCCESSFUL_SELF_ATTACK
+
+/datum/component/ability/devour/proc/can_devour_grabbed(atom/movable/AM)
+	var/mob/P = parent
```
Comment (ninjanomnom) [inline code/datums/components/abilities/component_devour.dm]: Use a real var name

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 7]
```diff
@@ -14,7 +14,7 @@
 		return
 	X.lay_down()
 	return succeed_activate()
-
+/*
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm]: Remove the code

**Pull request closed without merging**